File: oop-intro/code/rb_inheritance_simple_circle/main1.rb
#!/usr/bin/ruby
# encoding: utf-8

class Point
  attr_reader :x, :y

  ORIGIN = 0, 0

  def initialize(x, y)
    @x, @y = x, y
  end

  def >(other)
    @x > other.x
  end

  def <(other)
    @x < other.x
  end

  def >=(other)
    @x > other.x or @x == other.x
  end

  def <=(other)
    @x < other.x or @x == other.x
  end

  def ==(other)
    @x == other.x and @y == other.y
  end

  def +(other=nil)
    other ||= origin
    self.class.new @x + other.x, @y + other.y
  end

  def move(x, y)
    @x += x
    @y += y
    self
  end

  def to_s
    "(#@x,#{@y})"
  end

  # Sınıf metodları

  class << self
    def origin
      new *ORIGIN
    end

    def distance(here, there=nil)
      there ||= origin
      Math.sqrt((here.x - there.x)**2 + (here.y - there.y)**2)
    end
  end
end

class Circle < Point
  attr_reader :r

  def initialize(*coords, r)
    super *coords # XXX z,y yok; bu kısım artık DRY
    @r = r
  end

  def perimeter
    2 * Math::PI * @r
  end

  def area
    Math::PI * @r**2
  end

  # FIXME DRY ama ekstra satır
  alias :super_to_s :to_s
  def to_s
    "#@r@#{super_to_s}"
  end

  # FIXME Point'te overload edilen bazı işleçler
  # örneğin +, ==, çocuk sınıfta anlamlı değil.
end

def main
  c = Circle.new 3, 5, 2
  puts "#{c} çemberinin alanı: #{c.area}"
  puts "#{c} çemberi #{c.move(7, 9)} çemberine dönüştü"
end

if __FILE__ == $0
    main
end

File: oop-intro/code/py_inheritance_simple_circle/main1.py
#!/usr/bin/python
# encoding: utf-8

import math

class Point(object):
    def __init__(self, x, y):
        self.x, self.y = x, y

    def __eq__(self, other):
        return self.x is other.x and self.y is other.y
    def __lt__(self, other):
        return self.x < other.x
    def __gt__(self, other):
        return self.x > other.x
    def __le__(self, other):
        return self < other or self == other
    def __ge__(self, other):
        return self > other or self == other

    def __str__(self):
        return "(%s,%s)" % (self.x, self.y)

    def move(self, x, y):
        self.x += x
        self.y += y

class Circle(Point):
    def __init__(self, x, y, r):
        Point.__init__(self, x, y)
        self.r = r
    def __str__(self):
        return "%s@%s" % (self.r, Point.__str__(self))

    def perimeter(self):
        return 2 * math.pi * self.r

    def area(self):
        return math.pi * self.r**2

def main():
    c = Circle(3, 5, 2)
    print "%s çemberinin alanı: %f " % (c, c.area())
    print "%s çemberi " % c,
    c.move(7, 9)
    print "%s çemberine dönüştü." % c

if __name__ == "__main__":
    main()
